core/src/main/scala/cats/data/HashMap.scala

```diff
@@ -0,0 +1,1196 @@
+            i += 1
+          }
+          new BitMapNode[K, V](bitPos, 0, newContentsArray, newContents.size)
+        }
+      }
+
+    final def ===[VV >: V](that: Node[K, VV])(implicit eqValue: Eq[VV]): Boolean = {
+      (this eq that) || {
+        that match {
+          case node: CollisionNode[_, _] =>
+            (this.collisionHash === node.collisionHash) &&
+            (this.contents.size === node.contents.size) &&
+            this.contents.forall { case (kl, vl) =>
+              node.contents.exists { case (kr, vr) => hashKey.eqv(kl, kr) && eqValue.eqv(vl, vr) }
+            }
+          case _ =>
+            false
+        }
+      }
+    }
+
+    final override def equals(that: Any): Boolean = that match {
+      case node: CollisionNode[_, _] =>
+        (this.collisionHash == node.collisionHash) &&
+        (this.contents.size == node.contents.size) &&
+        this.contents.forall(node.contents.contains)
+      case _ =>
+        false
+    }
+
+    final override def toString(): String = {
+      s"""CollisionNode(hash=${collisionHash}, values=${contents.mkString("[", ",", "]")})"""
+    }
+  }
+
+  /**
+    * A CHAMP bitmap node. Stores value element and node element positions in the `contents` array
+    * in the `valueMap` and `nodeMap` integer bitmaps.
+    *
+    * @tparam A the type of the elements contained in this node.
+    * @param valueMap integer bitmap indicating the notional positions of value elements in the `contents` array.
+    * @param nodeMap integer bitmap indicating the notional positions of node elements in the `contents` array.
+    * @param contents an array of `A` value elements and `Node[A]` sub-node elements.
+    * @param size the number of value elements in this subtree.
+    */
+  final private[HashMap] class BitMapNode[K, +V](
+    val valueMap: Int,
+    val nodeMap: Int,
+    val contents: Array[Any],
+    val size: Int
+  )(implicit hashKey: Hash[K])
+      extends Node[K, V] {
+
+    final def hasValues: Boolean =
+      valueMap != 0
+
+    final def hasNodes: Boolean =
+      nodeMap != 0
+
+    final def allElements: Int =
+      valueElements + nodeElements
+
+    final def valueElements: Int =
+      Integer.bitCount(valueMap)
+
+    final def nodeElements: Int =
+      Integer.bitCount(nodeMap)
+
+    final private def hasNodeAt(bitPos: Int): Boolean =
+      (nodeMap & bitPos) != 0
+
+    final private def hasValueAt(bitPos: Int): Boolean =
+      (valueMap & bitPos) != 0
+
+    final def getKey(index: Int): K =
+      contents(Node.StrideLength * index).asInstanceOf[K]
+
+    final def getValue(index: Int): V =
+      contents(Node.StrideLength * index + 1).asInstanceOf[V]
+
+    final def getMapping(index: Int): (K, V) =
+      contents(Node.StrideLength * index).asInstanceOf[K] ->
+        contents(Node.StrideLength * index + 1).asInstanceOf[V]
+
+    final def getNode(index: Int): Node[K, V] =
+      contents(contents.length - 1 - index).asInstanceOf[Node[K, V]]
```

Comment:
this is some tricky stuff! `contents` has a very complex type: some offsets have `K`, some `V` and some `Node[K, V]`.

Reply:
I guess the front half is K, V pairs and the back half are the nodes, and the are reflected about the middle?